src/v/kafka/server/group_data_parser.h

```diff
@@ -0,0 +1,199 @@
+    vassert(
+      hdr.type == batch_type,
+      "broken tx group message. expected batch type {} got: {}",
+      hdr.type,
+      batch_type);
+    auto p_id = model::producer_id(reflection::adl<int64_t>{}.from(key_reader));
+    auto bid = model::batch_identity::from(hdr);
+    vassert(
+      p_id == bid.pid.id,
+      "broken tx group message. expected pid/id {} got: {}",
+      bid.pid.id,
+      p_id);
+    return cmd;
+}
+
+namespace kafka {
+
+template<class T>
+concept GroupDataParserBase = requires(T base, model::record_batch b) {
+    base.handle_raft_data(std::move(b));
+    base.handle_tx_offsets(b.header(), kafka::group_tx::offsets_metadata{});
+    base.handle_commit(b.header(), group_tx::commit_metadata{});
+    base.handle_abort(b.header(), group_tx::abort_metadata{});
+    base.handle_fence_v0(b.header(), group_tx::fence_metadata_v0{});
+    base.handle_fence_v1(b.header(), group_tx::fence_metadata_v1{});
+    base.handle_fence(b.header(), kafka::group_tx::fence_metadata{});
+    base.handle_version_fence(features::feature_table::version_fence{});
+};
+
+template<class Base>
+class group_data_parser {
+public:
+    group_data_parser() {
+        static_assert(
+          GroupDataParserBase<Base>,
+          "Base does not implement all the required methods.");
+    }
+
+protected:
+    ss::future<> parse(model::record_batch b) {
+        if (b.header().type == model::record_batch_type::raft_data) {
+            return handle_raft_data(std::move(b));
+        }
+        if (b.header().type == model::record_batch_type::group_prepare_tx) {
+            auto data = parse_tx_batch<kafka::group_tx::offsets_metadata>(
+              b, group::prepared_tx_record_version);
+            return handle_tx_offsets(b.header(), std::move(data));
+        }
+        if (b.header().type == model::record_batch_type::group_commit_tx) {
+            auto data = parse_tx_batch<group_tx::commit_metadata>(
+              b, group::commit_tx_record_version);
+            return handle_commit(b.header(), std::move(data));
+        }
+        if (b.header().type == model::record_batch_type::group_abort_tx) {
+            auto data = parse_tx_batch<group_tx::abort_metadata>(
+              b, group::aborted_tx_record_version);
+            return handle_abort(b.header(), std::move(data));
+        }
+        if (
+          b.header().type == model::record_batch_type::tx_fence
+          || b.header().type == model::record_batch_type::group_fence_tx) {
+            return parse_fence(std::move(b));
+        }
+        if (b.header().type == model::record_batch_type::version_fence) {
+            auto fence = features::feature_table::decode_version_fence(
+              std::move(b));
+            return handle_version_fence(fence);
+        }
+        vlog(klog.warn, "ignoring batch with type: {}", b.header().type);
+        return ss::make_ready_future<>();
+    }
+
+private:
+    ss::future<> parse_fence(model::record_batch b) {
```

Comment:
would probably be worth trying to move some of these implementations out of the header and into .cc files.